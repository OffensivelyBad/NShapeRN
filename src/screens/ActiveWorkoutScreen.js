'use strict';
import React from 'react';
import { Image, SafeAreaView, View, Text, Easing, Dimensions } from 'react-native';
import { AnimatedCircularProgress } from 'react-native-circular-progress';
import { BaseButton, TransparentCard } from '../components/common';
import * as Colors from '../components/Theme';
import Workout from '../models/Workout';
import Exercise from '../models/Exercise';

const pauseImagePath = '../images/activeWorkout/pause.png';
const playImagePath = '../images/activeWorkout/play.png';
const rewindImagePath = '../images/activeWorkout/rewind.png';
const fastforwardImagePath = '../images/activeWorkout/fastforward.png';
const screenWidth = Dimensions.get('window').width;

class ActiveWorkoutScreen extends React.PureComponent {

  isTimerRunning = false;
  isExercisePaused = true;
  timerFill = 100;

  constructor(props) {
    super(props);

    const newExercise1 = new Exercise('Push up', 5, 1);
    const newExercise2 = new Exercise('Pull up', 6, 5);
    const newExercise3 = new Exercise('Squat', 7, 4);
    const newExercise4 = new Exercise('Deadlift', 8, 3);
    const newExercise5 = new Exercise('Bench Press', 9, 2);
    const newExercise6 = new Exercise('Hoodilly Crushers', 10, 6);

    const exercises = [newExercise1] //, newExercise2, newExercise3, newExercise4, newExercise5, newExercise6];
    const duration = exercises.reduce((accumulator, currentValue) => {
      return accumulator + currentValue;
    });

    const newWorkout = new Workout('Test1', exercises, duration);

    const endWorkout = new Exercise('Complete!', 0, newWorkout.exercises.length + 1);
    newWorkout.exercises.push(endWorkout);

    this.state = {
      exercises: newWorkout.exercises,
      currentExerciseIndex: 0,
      currentExercise: {},
      playButtonImage: playImagePath
    }
  }

  componentDidMount() {
    const exercises = this._getSortedExercises(this.state.exercises);
    const { currentExerciseIndex } = this.state;
    const currentExercise = exercises.length > currentExerciseIndex ? exercises[currentExerciseIndex] : {}
    this.setState({ exercises, currentExercise });
  }

  _getSortedExercises = (exercises) => {
    const sortedData = exercises.sort((a, b) => {
      return (a.order < b.order) ? -1 : (a.order > b.order) ? 1 : 0;
    });
    return sortedData;
  }

  _getExercise = (exercises, index) => {
    if (!exercises.length) {
      // TODO: handle no exercises
      console.log('fail');
      return;
    }

    if (index < 0) {
      index = 0;
    }
    
    let exerciseIndex = 0;
    let exercise = exercises[exerciseIndex];
    if (exercises.length > index) {
      exerciseIndex = index;
      exercise = exercises[exerciseIndex];
    } else {
      // This is the end of the road
      exerciseIndex = -1;
    }
    return {exercise, exerciseIndex};
  }

  _endWorkout = () => {
    this.setState({ playButtonImage: playImagePath }, () => {
      this.isTimerRunning = false;
      this.isExercisePaused = false;
      this.timerFill = 100;
    });
  }

  _onPlayButtonPress = () => {
    if (this.isTimerRunning) {
      this._pauseExercise();
    } else {
      this.setState({ playButtonImage: pauseImagePath });
      let {exercise, exerciseIndex} = this._getExercise(this.state.exercises, this.state.currentExerciseIndex);
      if (exerciseIndex === -1) {
        // TODO: end the workout?
        this._endWorkout();
      } else {
        if (this.isExercisePaused) {
          this._resumeExercise(exercise, this.timerFill);
        } else {
          this.isTimerRunning = true;
          this._startExercise(exercise);
        }
      }
    }
  }

  _onPreviousButtonPress = () => {
    this._pauseExercise();
    this._onExerciseEnd(true);
  }

  _onNextButtonPress = () => {
    this._pauseExercise();
    this._onExerciseEnd();
  }

  _pauseExercise = () => {
    if (!this.isTimerRunning || this.isExercisePaused) {
      return;
    }
    this.isTimerRunning = false;
    this.isExercisePaused = true;
    this.circularProgress.pause();
    this.setState({ playButtonImage: playImagePath });
  } 

  _resumeExercise = (exercise, remainingPercent) => {
    const remainingTime = (exercise.duration * (remainingPercent / 100)) * 1000;
    this.circularProgress.animate(1, remainingTime, Easing.linear);
    this.isExercisePaused = false;
    this.isTimerRunning = true;
  }

  _startExercise = (exercise) => {
    this.isTimerRunning = true;
    this.isExercisePaused = false;
    this.circularProgress.reAnimate(100, 1, exercise.duration * 1000, Easing.linear);
  }

  _onExerciseEnd = (goToPrevious) => {
    const nextIndex = goToPrevious ? this.state.currentExerciseIndex - 1 : this.state.currentExerciseIndex + 1;
    const {exercise, exerciseIndex} = this._getExercise(this.state.exercises, nextIndex);
    if (exerciseIndex == -1) {
      this._endWorkout();
    } else {
      console.log(`Starting exerciseIndex: ${exerciseIndex} and exercise: ${exercise}`);
      this._setNextExercise(exerciseIndex, exercise);
    }
  }

  _setNextExercise = (index, exercise) => {
    this.setState({ currentExerciseIndex: index, currentExercise: exercise }, () => {
      if (!this.isExercisePaused) { 
        this._startExercise(exercise);
      }
    });
  }

  _onAnimationEnd = () => {
    if (this.isTimerRunning) {
      this.isTimerRunning = false;
      this._onExerciseEnd();
    }
  }

  _getTimerString = (duration, fillPercent) => {
    const timeRemaining = Math.round((duration * fillPercent) / 100);
    let minutes = 0;
    let seconds = timeRemaining;
    if (timeRemaining > 60) {
      minutes = timeRemaining / 60;
      seconds = timeRemaining % 60;
    }

    let formattedTime = minutes + ':' + seconds;

    if (seconds < 10) {
      formattedTime = minutes + ':0' + seconds;
    }
    
    return formattedTime;
  }

  _renderProgress = () => {
    let { exercise, exerciseIndex } = this._getExercise(this.state.exercises, this.state.currentExerciseIndex);

    if (exerciseIndex === -1) {
      // TODO: end the workout?
      this._endWorkout();
    }

    return (
      <AnimatedCircularProgress
        ref={(ref) => this.circularProgress = ref}
        fill={100}
        rotation={-360}
        size={200}
        width={4}
        tintColor='red'
        onAnimationComplete={this._onAnimationEnd}
      >
      {(fill) => {
        this.timerFill = fill;
        return (
        <Text style={[styles.timerTextStyle, {color:Colors.DarkGray}]}>
          { this._getTimerString(exercise.duration, fill) }
        </Text>
      )}}
      </AnimatedCircularProgress>
    )
  }

  _renderButtonControls = () => {
    return (
      <View style={styles.controlsContainerStyle}>
        <BaseButton style={styles.controlButtonStyle} onPress={this._onPreviousButtonPress}>
          <Image 
            source={require(rewindImagePath)} 
            style={styles.controlButtonImageStyle} 
            resizeMode='contain' 
          />
        </BaseButton>
        <BaseButton style={styles.controlButtonStyle} onPress={this._onPlayButtonPress}>
          {this._renderPausePlayButton(this.isExercisePaused, this.isTimerRunning)}
        </BaseButton>
        <BaseButton style={styles.controlButtonStyle} onPress={this._onNextButtonPress}>
          <Image 
            source={require(fastforwardImagePath)} 
            style={styles.controlButtonImageStyle} 
            resizeMode='contain' 
          />
          </BaseButton>
      </View>
    );
  }

  _renderPausePlayButton = (isExercisePaused, isTimerRunning) => {
    if (isExercisePaused || !isTimerRunning) {
      return (
        <Image 
          source={require(playImagePath)} 
          style={styles.controlButtonImageStyle} 
          resizeMode='contain' 
        />
      );
    } else {
      return (
        <Image 
          source={require(pauseImagePath)} 
          style={styles.controlButtonImageStyle} 
          resizeMode='contain' 
        />
      );
    }
  }

  _renderExerciseName = (name) => {
    return (
      <TransparentCard style={styles.titleContainerStyle}>
        <Text style={styles.titleTextStyle}>{name}</Text>
      </TransparentCard>
    );
  }

  render() {
    return (
      <SafeAreaView style={styles.containerStyle}>
        <View style={styles.exercisesContainerStyle}>
          {this._renderExerciseName(this.state.currentExercise.name)}
        </View>
        <View style={styles.clockContainerStyle}>
          {this._renderProgress()}
        </View>
        {this._renderButtonControls()}
      </SafeAreaView>
    )
  }
}

const styles = {
  containerStyle: {
    flex: 1,
    alignItems: 'center',
  },
  clockContainerStyle: {
    flex: 4,
    justifyContent: 'flex-end'
  },
  exercisesContainerStyle: {
    flex: 5,
    justifyContent: 'center'
  },
  exerciseTextStyle: {
    fontSize: 49,
  },
  controlsContainerStyle: {
    flex: 4,
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center'
  },
  timerTextStyle: {
    fontSize: 49,
  },
  controlButtonStyle: {
    alignItems: 'center',
    alignSelf: 'flex-end',
    marginBottom: 20
  },
  controlButtonImageStyle: {
    width: 50,
    height: 50,
  },
  titleContainerStyle: {
    position: 'absolute',
    justifyContent: 'center',
    alignSelf: 'center',
    width: screenWidth - 100,
    height: 100,
    backgroundColor: Colors.ThemeColor
  },
  titleTextStyle: {
    alignSelf: 'center',
    fontSize: 49,
    color: '#ffffff'
  }
}

export default ActiveWorkoutScreen;
