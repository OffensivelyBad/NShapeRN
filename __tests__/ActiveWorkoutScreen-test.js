import React from 'react';
import ActiveWorkoutScreen from '../src/screens/ActiveWorkoutScreen';
import renderer from 'react-test-renderer';

test('activeworkoutscreen renders correctly', () => {
    const tree = renderer.create(<ActiveWorkoutScreen />).toJSON();
    expect(tree).toMatchSnapshot();
});
