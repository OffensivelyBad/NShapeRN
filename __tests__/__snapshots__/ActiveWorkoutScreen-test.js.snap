// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`activeworkoutscreen renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "backgroundColor": "#FFFFFF",
            "borderRadius": 12,
            "flex": 1,
            "opacity": 0.8,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 2,
              "width": 0,
            },
            "shadowOpacity": 0.7,
            "shadowRadius": 2,
          },
          Object {
            "alignSelf": "center",
            "backgroundColor": "#f10026",
            "height": 100,
            "justifyContent": "center",
            "position": "absolute",
            "width": 650,
          },
        ]
      }
    >
      <Text
        style={
          Object {
            "alignSelf": "center",
            "color": "#ffffff",
            "fontSize": 49,
          }
        }
      >
        Push up
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View>
      <RNSVGSvgView
        bbHeight={200}
        bbWidth={200}
        height={200}
        style={
          Array [
            Object {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            Object {
              "backgroundColor": "transparent",
            },
            null,
            Object {
              "flex": 0,
              "height": 200,
              "width": 200,
            },
          ]
        }
        width={200}
      >
        <RNSVGGroup
          fill={
            Array [
              0,
              4278190080,
            ]
          }
          fillOpacity={1}
          fillRule={1}
          font={Object {}}
          matrix={
            Array [
              1,
              0,
              0,
              1,
              0,
              0,
            ]
          }
          opacity={1}
          originX={0}
          originY={0}
          propList={
            Array [
              "fill",
              "fillOpacity",
              "fillRule",
              "stroke",
              "strokeWidth",
              "strokeOpacity",
              "strokeDasharray",
              "strokeDashoffset",
              "strokeLinecap",
              "strokeLinejoin",
              "strokeMiterlimit",
            ]
          }
          rotation={0}
          scaleX={1}
          scaleY={1}
          skewX={0}
          skewY={0}
          stroke={null}
          strokeDasharray={null}
          strokeDashoffset={null}
          strokeLinecap={0}
          strokeLinejoin={0}
          strokeMiterlimit={4}
          strokeOpacity={1}
          strokeWidth={1}
          x={0}
          y={0}
        >
          <RNSVGGroup
            fill={
              Array [
                0,
                4278190080,
              ]
            }
            fillOpacity={1}
            fillRule={1}
            font={Object {}}
            matrix={
              Array [
                1,
                0,
                -0,
                1,
                0,
                0,
              ]
            }
            opacity={1}
            originX={100}
            originY={100}
            propList={Array []}
            rotation={-360}
            scaleX={1}
            scaleY={1}
            skewX={0}
            skewY={0}
            stroke={null}
            strokeDasharray={null}
            strokeDashoffset={null}
            strokeLinecap={0}
            strokeLinejoin={0}
            strokeMiterlimit={4}
            strokeOpacity={1}
            strokeWidth={1}
            x={0}
            y={0}
          />
        </RNSVGGroup>
      </RNSVGSvgView>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderRadius": 96,
            "height": 192,
            "justifyContent": "center",
            "left": 4,
            "overflow": "hidden",
            "position": "absolute",
            "top": 4,
            "width": 192,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontSize": 49,
              },
              Object {
                "color": "#4A4A4A",
              },
            ]
          }
        >
          0:00
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": undefined,
          "borderRadius": 5,
          "flex": 1,
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/images/activeWorkout/rewind.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": undefined,
          "borderRadius": 5,
          "flex": 1,
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/images/activeWorkout/play.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
    <View
      accessible={true}
      isTVSelectable={true}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "alignSelf": "stretch",
          "backgroundColor": undefined,
          "borderRadius": 5,
          "flex": 1,
          "height": 50,
          "justifyContent": "center",
          "marginLeft": 10,
          "marginRight": 10,
          "opacity": 1,
          "shadowColor": undefined,
          "shadowOffset": undefined,
          "shadowOpacity": undefined,
          "shadowRadius": undefined,
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/images/activeWorkout/fastforward.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 50,
            "width": 50,
          }
        }
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
